moveit_setup_assistant/src/widgets/setup_assistant_widget.cpp

```diff
@@ -237,118 +227,66 @@ void SetupAssistantWidget::moveToScreen(const int index)
-  connect(controllers_widget_, SIGNAL(highlightLink(const std::string&, const QColor&)), this,
-          SLOT(highlightLink(const std::string&, const QColor&)));
-  connect(controllers_widget_, SIGNAL(highlightGroup(const std::string&)), this,
-          SLOT(highlightGroup(const std::string&)));
-  connect(controllers_widget_, SIGNAL(unhighlightAll()), this, SLOT(unhighlightAll()));
-
-  // Simulation Screen
-  simulation_widget_ = new SimulationWidget(this, config_data_);
-  main_content_->addWidget(simulation_widget_);
-  connect(simulation_widget_, SIGNAL(isModal(bool)), this, SLOT(setModalMode(bool)));
-  connect(simulation_widget_, SIGNAL(highlightLink(const std::string&, const QColor&)), this,
-          SLOT(highlightLink(const std::string&, const QColor&)));
-  connect(simulation_widget_, SIGNAL(highlightGroup(const std::string&)), this,
-          SLOT(highlightGroup(const std::string&)));
-  connect(simulation_widget_, SIGNAL(unhighlightAll()), this, SLOT(unhighlightAll()));
-
-  // Perception
-  perception_widget_ = new PerceptionWidget(this, config_data_);
-  main_content_->addWidget(perception_widget_);
-  connect(perception_widget_, SIGNAL(isModal(bool)), this, SLOT(setModalMode(bool)));
-  connect(perception_widget_, SIGNAL(highlightLink(const std::string&, const QColor&)), this,
-          SLOT(highlightLink(const std::string&, const QColor&)));
-  connect(perception_widget_, SIGNAL(highlightGroup(const std::string&)), this,
-          SLOT(highlightGroup(const std::string&)));
-  connect(perception_widget_, SIGNAL(unhighlightAll()), this, SLOT(unhighlightAll()));
-
-  // Author Information
-  author_information_widget_ = new AuthorInformationWidget(this, config_data_);
-  main_content_->addWidget(author_information_widget_);
-  connect(author_information_widget_, SIGNAL(isModal(bool)), this, SLOT(setModalMode(bool)));
-  connect(author_information_widget_, SIGNAL(highlightLink(const std::string&, const QColor&)), this,
-          SLOT(highlightLink(const std::string&, const QColor&)));
-  connect(author_information_widget_, SIGNAL(highlightGroup(const std::string&)), this,
-          SLOT(highlightGroup(const std::string&)));
-  connect(author_information_widget_, SIGNAL(unhighlightAll()), this, SLOT(unhighlightAll()));
-
-  // Configuration Files
-  configuration_files_widget_ = new ConfigurationFilesWidget(this, config_data_);
-  main_content_->addWidget(configuration_files_widget_);
+  if (!plugins_loader_)
+  {
+    try
+    {
+      plugins_loader_.reset(
+          new pluginlib::ClassLoader<SetupScreenWidget>("moveit_setup_assistant", "SetupScreenWidget"));
+    }
+    catch (pluginlib::PluginlibException& ex)
+    {
+      ROS_FATAL_STREAM("Exception while creating octomap updater plugin loader " << ex.what());
+    }
+  }
+
+  // Declared classes with setup screen widget as their base class
+  std::vector<std::string> classes = plugins_loader_->getDeclaredClasses();
+
+  SetupScreenWidget* widget;
```

Comment:
is this a memory leak? you don't track the created object so it might not be destroyed when the class deconstructs

Reply:
Okay I remember suspecting this part is causing a memory leak too, but according to [qt doc](https://doc.qt.io/qt-5/objecttrees.html) , this widget is automatically deleted when its parent is deleted .... still maybe I'm wrong.

Reply:
ah right, this line takes ownership of the memory:

>   QStackedLayout* main_content_;

resolved.